keps/sig-apps/4040-controller-name/README.md

```diff
@@ -0,0 +1,357 @@
+
+## Motivation
+
+Currently Kubernetes users, who need custom behavior for
+some API objects, can either:
+
+* disable the corresponding controller in controller-manager
+and implement a new one that provides both modified(for marked
+objects) and original(for all other) functionality.
+
+* duplicate/fork the API as CRD.
+
+However, on managed environments (like, for example, GKE),
+only the second option is available as cloud providers often
+don't allow to modify `controller-manager` configuration.
+
+### Goals
+
+* Establish a simple and common method to mark an API object as
+to be handled by a non default controller, in a similar fashion
+as pods can specify their custom scheduler.
+
+* Apply it and provide implementation for Job and HPA.
+
+### Non-Goals
+
+* Add controller name to all possible K8S objects at once.
+
+
+* While the KEP enables providing custom handling for standard
+K8S APIs,
+it doesn't focus on any customization or additional
+configuration that might be possible using it. Additional
+configuration options may be subject for the following
+enhancement proposals.
+
+
+## Proposal
+
+Add `controllerName` at the top level of Spec of API objects to
+specify which controller should handle them. If the field
+value is `”default”` or the field is not specified at all,
+then the default controller from controller manager handles the
+object in the usual way. If the specified  value is different,
+then the controller, which identifies itself with the given name, should handle the object and all others ignore it.
+
+Initially add the field only to Job and HPA. If the feedback is good, consider expansion to other objects.
+
+### User Stories (Optional)
+
+#### Story 1
+
+I want to run a modified, custom version of  ControllerManager in
+cluster and handle some (not all) of the Jobs with it.
+
+#### Story 2
+
+I want to provide a significantly different implementation of
+HPA and gradually test it in production.
+
+### Notes/Constraints/Caveats (Optional)
+
+Other controller owners may decide whether and when to introduce
+the field. Allowing it on security-related APIs may be more
+controversial than on `HorizontalPodAutoscaler` or `Job`.
+
+### Risks and Mitigations
+
+Objects handled by custom controllers may have different field value
+combinations than handled by the default controller manager. These
+combinations, while correct from the API validation perspective,
+may project logical error or simply be not understood by dependant
+code, and, as a result, potentially crash some external components.
+
+Adding the field on security-related APIs may potentially
+mislead the system administrators and, as a result misconfiguration risks.
+
+To minimize the risks we can start with supporting the field
+on just 2 APIs - HPA and Job, but the same field can be added to other objects.
+
+## Design Details
+
+The supported API controllers will simply skip API objects that have non-default `controllerName` field.
```

Comment:
What about defaulting from apiserver?

Would it set the value to `default`?

One lesson learned from `schedulerName` is that we shouldn't do it at the defaulting layer. We should do it in the strategy https://github.com/kubernetes/kubernetes/issues/93067

Reply:
Added.